In [5]:
import numpy as np
import pandas as pd
from thefuzz import process

In [12]:
# Load Data
df = pd.read_excel("../data/raw/selected_indicators_data.xlsx")

In [20]:
# Basic Check
print("Shape : \n",df.shape)
print("\n Columns : \n",df.columns.tolist())

Shape : 
 (707, 21)

 Columns : 
 ['District Names', 'State/UT', 'Women (age 15-49 years) whose Body Mass Index (BMI) is below normal (BMI <18.5 kg/m2)21 (%)', 'Women (age 15-49 years) who are overweight or obese (BMI ≥25.0 kg/m2)21 (%)', 'All women age 15-49 years who are anaemic22 (%)', 'Women age 15 years and above wih high or very high (>140 mg/dl) Blood sugar level or taking medicine to control blood sugar level23 (%)', 'Men age 15 years and above wih high or very high (>140 mg/dl) Blood sugar level  or taking medicine to control blood sugar level23 (%)', 'Women age 15 years and above wih Elevated blood pressure (Systolic ≥140 mm of Hg and/or Diastolic ≥90 mm of Hg) or taking medicine to control blood pressure (%)', 'Men age 15 years and above wih Elevated blood pressure (Systolic ≥140 mm of Hg and/or Diastolic ≥90 mm of Hg) or taking medicine to control blood pressure (%)', 'Women (age 30-49 years) Ever undergone a screening test for cervical cancer (%)', 'Women (age 30-49 years)

In [ ]:
print("Null Values : \n",df.isnull().sum())


 Null Values : 
 District Names                                                                                                                                                      0
State/UT                                                                                                                                                            0
Women (age 15-49 years) whose Body Mass Index (BMI) is below normal (BMI <18.5 kg/m2)21 (%)                                                                         0
Women (age 15-49 years) who are overweight or obese (BMI ≥25.0 kg/m2)21 (%)                                                                                         0
All women age 15-49 years who are anaemic22 (%)                                                                                                                     0
Women age 15 years and above wih high or very high (>140 mg/dl) Blood sugar level or taking medicine to control blood sugar level23 (%)                 

In [ ]:
print("Data Types : \n",df.dtypes)


 Data Types : 
 District Names                                                                                                                                                       object
State/UT                                                                                                                                                             object
Women (age 15-49 years) whose Body Mass Index (BMI) is below normal (BMI <18.5 kg/m2)21 (%)                                                                         float64
Women (age 15-49 years) who are overweight or obese (BMI ≥25.0 kg/m2)21 (%)                                                                                         float64
All women age 15-49 years who are anaemic22 (%)                                                                                                                     float64
Women age 15 years and above wih high or very high (>140 mg/dl) Blood sugar level or taking medicine to control blood sugar

In [42]:
# Statistical summary
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
print(df.describe().round(2).T)

                      count   mean    std    min    25%    50%    75%     max
bmi_below_normal_w    707.0  17.89   7.45   1.17  12.66  17.99  23.08   43.58
overweight_obese_w    707.0  23.07  10.60   3.77  14.88  21.26  30.18   53.01
anaemia_w             707.0  55.94  11.97  14.91  49.25  57.18  63.66   93.50
blood_sugar_w         707.0  12.46   4.62   4.06   9.00  11.67  15.26   32.12
blood_sugar_m         707.0  14.73   4.95   4.79  11.11  14.11  17.66   36.24
bp_elevated_w         707.0  21.40   5.22   8.53  17.62  21.01  24.46   42.09
bp_elevated_m         707.0  24.76   6.77  10.02  19.82  24.42  28.97   49.61
cervical_screening_w  707.0   1.57   2.77   0.00   0.19   0.55   1.52   23.22
breast_exam_w         707.0   0.65   1.57   0.00   0.00   0.21   0.50   14.55
oral_exam_w           707.0   0.70   1.47   0.00   0.00   0.28   0.69   15.84
tobacco_w             707.0  11.61  11.94   0.06   4.09   7.67  14.76   70.58
tobacco_m             707.0  40.60  14.08   6.75  30.42  42.49  

In [43]:
# Column rename
df.columns = [
    'district', 'state',
    'bmi_below_normal_w', 'overweight_obese_w',
    'anaemia_w',
    'blood_sugar_w', 'blood_sugar_m',
    'bp_elevated_w', 'bp_elevated_m',
    'cervical_screening_w', 'breast_exam_w', 'oral_exam_w',
    'tobacco_w', 'tobacco_m',
    'alcohol_w', 'alcohol_m',
    'literacy_w',
    'health_insurance',
    'clean_water',
    'sanitation',
    'clean_fuel'
]

print(df.columns.tolist())
print("Shape:", df.shape)

['district', 'state', 'bmi_below_normal_w', 'overweight_obese_w', 'anaemia_w', 'blood_sugar_w', 'blood_sugar_m', 'bp_elevated_w', 'bp_elevated_m', 'cervical_screening_w', 'breast_exam_w', 'oral_exam_w', 'tobacco_w', 'tobacco_m', 'alcohol_w', 'alcohol_m', 'literacy_w', 'health_insurance', 'clean_water', 'sanitation', 'clean_fuel']
Shape: (707, 21)


In [44]:
# District and State name cleaning
df['district'] = df['district'].str.strip().str.title()
df['state'] = df['state'].str.strip().str.title()

# Check
print("Sample districts:")
print(df['district'].head(10).tolist())
print("\nUnique states:", df['state'].nunique())
print(df['state'].unique())

Sample districts:
['Nicobars', 'North & Middle Andaman', 'South Andaman', 'Srikakulam', 'Vizianagaram', 'Visakhapatnam', 'East Godavari', 'West Godavari', 'Krishna', 'Guntur']

Unique states: 36
['Andaman & Nicobar Islands' 'Andhra Pradesh' 'Arunachal Pradesh' 'Assam'
 'Bihar' 'Chandigarh' 'Chhattisgarh'
 'Dadra And Nagar Haveli & Daman And Diu' 'Goa' 'Gujarat' 'Haryana'
 'Himachal Pradesh' 'Jammu & Kashmir' 'Jharkhand' 'Karnataka' 'Kerala'
 'Ladakh' 'Lakshadweep' 'Madhya Pradesh' 'Maharastra' 'Manipur'
 'Meghalaya' 'Mizoram' 'Nagaland' 'Nct Of Delhi' 'Odisha' 'Puducherry'
 'Punjab' 'Rajasthan' 'Sikkim' 'Tamil Nadu' 'Telangana' 'Tripura'
 'Uttar Pradesh' 'Uttarakhand' 'West Bengal']


In [45]:
# Numeric columns values 0-100 range check
numeric_cols = df.columns[2:]

for col in numeric_cols:
    below = (df[col] < 0).sum()
    above = (df[col] > 100).sum()
    if below > 0 or above > 0:
        print(f"{col}: {below} below 0, {above} above 100")

print("Range check complete")

Range check complete


In [40]:
# Duplicate districts check
duplicates = df[df.duplicated(subset=['district', 'state'], keep=False)]
print(f"Duplicate rows: {len(duplicates)}")
if len(duplicates) > 0:
    print(duplicates[['district', 'state']])

Duplicate rows: 0


In [41]:
# Save cleaned data
df.to_excel('../data/processed/pulsemap_cleaned.xlsx', index=False)
print("Cleaned data saved successfully")
print("Shape:", df.shape)

Cleaned data saved successfully
Shape: (707, 21)


## Cleaning Summary
- Source: NFHS-5 District Factsheet (data.gov.in)
- Initial shape: (707, 21)
- Final shape: (707, 21)
- Null values: None found
- Duplicates: None found
- Range violations: None found
- Column names: Standardized to snake_case
- District/State names: Stripped and Title Cased